File: kinotic-orchestrator/src/main/java/org/kinotic/orchestrator/api/model/grind/StepInfo.java


package org.kinotic.orchestrator.api.model.grind;

/**
 * The sequence of steps that have been executed to get to a specific {@link Result}
 *
 * Created by Navid Mitchell on 11/18/20
 */
public class StepInfo {

    private final int sequence;

    private StepInfo ancestor = null;

    private StepInfo top = null;

    public StepInfo(int sequence) {
        this.sequence = sequence;
    }

    /**
     * The sequence of the step that created this
     * @return the sequence number
     */
    public int getSequence() {
        return sequence;
    }

    public void addAncestor(StepInfo ancestor){
        if(this.ancestor == null){
            this.ancestor = ancestor;
        }else{
            this.top.addAncestor(ancestor);
        }
        this.top = ancestor;
    }

    public StepInfo getAncestor() {
        return ancestor;
    }

    /**
     * The {@code /} separated sequence path from the run's root down to the step that created
     * this info, matching the {@code stepPath} recorded on a {@code TaskRecord}.
     * @return the step path, complete once the {@link Result} carrying this info has been
     *         delivered to a subscriber of the run
     */
    public String path() {
        StringBuilder ret = new StringBuilder();
        for(StepInfo info = this; info != null; info = info.getAncestor()){
            if(!ret.isEmpty()){
                ret.insert(0, '/');
            }
            ret.insert(0, info.getSequence());
        }
        return ret.toString();
    }
}

File: kinotic-js/workspace/packages/management-api/src/api/model/grind/StepInfo.ts
/**
 * The sequence of steps that were executed to get to a specific Result.
 */
export class StepInfo {

    /**
     * The sequence of the step that created the Result.
     */
    public sequence: number = 0

    /**
     * The enclosing step's info, or null for the run's root.
     */
    public ancestor: StepInfo | null = null

}

/**
 * The `/` separated sequence path from the run's root down to the step that created the
 * given info, matching the stepPath recorded on a TaskRecord.
 * @param info the step info of a received Result
 * @return the step path, e.g. "0/2/1"
 */
export function stepPathOf(info: StepInfo): string {
    let ret = ''
    for (let current: StepInfo | null = info; current !== null; current = current.ancestor) {
        ret = ret.length > 0 ? `${current.sequence}/${ret}` : String(current.sequence)
    }
    return ret
}
